### ISSUE TITLE
Unable to save created curved geometries

### ISSUE THREAD

**Issue by jgrocha** [qgis/QGIS#39974]
### Support for CurvePolygon, CompoundCurve and CircularString

To test the support for these primitive geometry types, I have used the wedge buffer tool. I've created a point and then a wedge buffer around it (in memory). The geometry (wkt) created is:
```
CurvePolygon (CompoundCurve (CircularString (-23776 205602, -23772 205603, -23768 205602),(-23768 205602, -23770 205598),CircularString (-23770 205598, -23772 205598, -23774 205598),(-23774 205598, -23776 205602)))
```
The following print screen show this setup.
![wedge buffer](https://user-images.githubusercontent.com/163681/98855398-d4711b80-2453-11eb-911d-d033286293c2.png)

#### Store the geometry on the providers Postgis and Geopackage

By just dragging and drop, **I can store this wedge buffer using both the Geopackage and Postgis providers**. On both providers, the geometry is stored using this primitive geometries CurvePolygon, CompoundCurve and CircularString.

### Editing existing geometry works!

If I edit an existing geometry with `CurvePolygon` (change vertexes, for example) it can be saved back to the provider.

To reproduce:

1. Open a wedge buffer generated layer
2. Edit one geometry
3. Save the layer

Works both for Geopackage and Postgis.

### The problem

If I edit the layer and **create a new geometry using the curve tool**, I'm not able to store it. 

#### To reproduce using Geopackage

1. Create a layer in Geopackage, CurvePolygon type (or use and existent CurvePolygon with existent CurvePolygon geometries)
2. Create a feature, using the Curve tool
3. Save the layer

The provider return the error:
```
OGR error creating wkb for feature -29: Invalid curve: not enough points
```
#### To reproduce using Postgis

The same happens for the Postgis provider. 

1. Create a layer in Postgis, CurvePolygon type (or use and existent CurvePolygon with existent CurvePolygon geometries)
2. Create a feature, using the Curve tool
3. Save the layer

The error returned is:
```
  Provider errors:
      PostGIS error while adding features: ERROR:  LineString must have at least two points
    CONTEXT:  SQL function "st_geomfromwkb" statement 1
```
In the postgresql log, the error reported is:
```
LOG:  statement: BEGIN
ERROR:  LineString must have at least two points
CONTEXT:  SQL function "st_geomfromwkb" statement 1
STATEMENT:  INSERT INTO "public"."lotes_p_curve"("geom") VALUES (st_geomfromwkb($1::bytea,3763)) RETURNING "id"
statement: ROLLBACK
statement: DEALLOCATE addfeatures
(...)
```
### WKT of the created geometry

It fails to write the feature to the provider. But while it is in memory, and can check its wkt representation, with looks fine to me:
```
CurvePolygon (CompoundCurve ((-23798 205603),(-23798 205603, -23786 205603),CircularString (-23786 205603, -23790 205597, -23798 205595),(-23798 205595, -23798 205603)))
```
I'm using this tiny PyQGIS to dump the wkt representation:
```python
layer = iface.activeLayer()

for f in list(layer.getFeatures()):
    geom = f.geometry()
    print(geom.asWkt(0))
```
QGIS is able to parse this Wkt representation back to a geometry:
```python
c = QgsGeometry().fromWkt("CurvePolygon (CompoundCurve ((-23798 205603),(-23798 205603, -23786 205603),CircularString (-23786 205603, -23790 205597, -23798 205595),(-23798 205595, -23798 205603)))")
print(c)
print(QgsWkbTypes.displayString(c.wkbType()))
```
Result:
```
<QgsGeometry: CurvePolygon (CompoundCurve ((-23798 205603),(-23798 205603, -23786 205603),CircularString (-23786 205603, -23790 205597, -23798 205595),(-23798 205595, -23798 205603)))>
CurvePolygon
```

### QGIS version

QGIS version | 3.17.0-Master
Compiled against Qt | 5.12.8
Compiled against GDAL/OGR | 3.3.0dev-5a81e514d8-dirty
Compiled against GEOS | 3.8.0-CAPI-1.13.1
Compiled against SQLite | 3.31.1
PostgreSQL Client Version | 13.0 (Ubuntu 13.0-1.pgdg20.04+1)
QWT Version | 6.1.4
Compiled against PROJ | 7.2.0

**Comment by elpaso** [comment]
I can confirm that there is something wrong in `asWkb`, the failing tests are wkb related:

```
Totals: 59 passed, 6 failed, 0 skipped, 0 blacklisted, 1681ms
********* Finished testing of TestQgsGeometry *********
```

Since Travis is using GDAL 2 I guess nobody noticed.

**Comment by elpaso** [comment]
wait, I'm not yet sure about that: it might well be a digitizing issue after all, it seems like the created CurvePolygon is wrong, even `asWkt` returns a wierd output (see the first element in the curve):

```
CurvePolygon (CompoundCurve ((-1.44649446494464939 -0.20479704797047971),CircularString (-1.44649446494464939 -0.20479704797047971, -1.3745387453874538 0.34686346863468631, -0.80811808118081174 0.36900369003690037),(-0.80811808118081174 0.36900369003690037, -1.44649446494464939 -0.20479704797047971)))
```

**Comment by jgrocha** [comment]
## Digitizing issue
This seems to be a digitizing issue. Using PyQGIS to create and write features with curved geometries is working. It seems that the QGIS representation and `asWkb()` conversion is working. 

### Digitizing with curve ![mActionDigitizeWithCurve](https://user-images.githubusercontent.com/163681/99157800-31730880-26c4-11eb-876e-f9a8c668ad96.png)

When digitizing using **Digitize with curve**, the geometry has an additional point at the beginning. Removing this point works.

Generated geometry by digitize tool:

```
CurvePolygon (CompoundCurve ((-23817 205627),(-23817 205627, -23815 205633),(-23815 205633, -23809 205632),CircularString (-23809 205632, -23807 205629, -23810 205624),(-23810 205624, -23817 205627)))
```
Fixed geometry:
<pre>
CurvePolygon (CompoundCurve (<b><del>(-23817 205627),</del></b>(-23817 205627, -23815 205633),(-23815 205633, -23809 205632),CircularString (-23809 205632, -23807 205629, -23810 205624),(-23810 205624, -23817 205627)))
</pre>

As in any polygon, the first point must equals the last one. But the first point is duplicated. By eliminating the first point, we get a valid geometry.

This works (the digitized geometry with the first point removed:
```python
layer = iface.activeLayer()
feature5 = QgsFeature(layer.fields())
geom5 = QgsGeometry.fromWkt("CurvePolygon (CompoundCurve ((-23817 205627, -23815 205633),(-23815 205633, -23809 205632),CircularString (-23809 205632, -23807 205629, -23810 205624),(-23810 205624, -23817 205627)))")
print(geom5)
feature5.setGeometry(geom5)
layer.startEditing()
layer.dataProvider().addFeatures([feature5])
layer.commitChanges()
```

**Comment by jgrocha** [comment]
### The culprit: CompoundCurve

This issue is related with `CompoundCurve` generation, when digitizing.

I've created a simple CompondCurve based layer. Digitizing is producing invalid geometries, as described above (for polygons).

Invalid geometry, with an additional first point at the head of the `CompondCurve`:
```
CompoundCurve ((-23828 205609),(-23828 205609, -23823 205612),(-23823 205612, -23822 205607),(-23822 205607, -23819 205612),(-23819 205612, -23818 205607),CircularString (-23818 205607, -23816 205611, -23812 205612),(-23812 205612, -23811 205607),(-23811 205607, -23808 205611),CircularString (-23808 205611, -23806 205610, -23804 205606))
```
Removing the first point works:
```python
layer = iface.activeLayer()

feature = QgsFeature(layer.fields())
geom = QgsGeometry.fromWkt("CompoundCurve ((-23828 205609, -23823 205612),(-23823 205612, -23822 205607),(-23822 205607, -23819 205612),(-23819 205612, -23818 205607),CircularString (-23818 205607, -23816 205611, -23812 205612),(-23812 205612, -23811 205607),(-23811 205607, -23808 205611),CircularString (-23808 205611, -23806 205610, -23804 205606))")
print(geom)
feature.setGeometry(geom)

layer.startEditing()
layer.dataProvider().addFeatures([feature])
layer.commitChanges()
```

**Comment by jgrocha** [comment]
### Internal representation

As far as I can check, after digitizing, the feature is represented with an additional component. `asWkb()`, `asWkt()`, etc will fail afterwards. 

Digitizing the most simple curve, with just three points:
![digitized curve](https://user-images.githubusercontent.com/163681/99159062-1fe42d80-26d1-11eb-9fd8-61883e70499c.png)
will became internally a curve with 2 components.

By debugging `QgsCompoundCurve::asWkb`, I've got a list of size 2, with these components (shown as Wkt):
```
LineString (-23789 205617)
CircularString (-23789 205617, -23792 205632, -23774 205626)
```

**Comment by jgrocha** [comment]
This issue might have been introduced by https://github.com/qgis/QGIS/commit/932151f33a40620f0ee4192c6b987c8df8fa30e0
I've tested with its [parent commit](https://github.com/qgis/QGIS/commit/f3a8dd792a) and the digitizing works.

What do you think @vcloarec? Can you take a look?

**Comment by vcloarec** [comment]
@jgrocha , yes, it seems that the new tool digitising with curve add this one point line string at the beginning. I look at to fix this.
Have you seen this issue using other tools?

**Comment by vcloarec** [comment]
@jgrocha , hope that it is fixed by #40059. 
Thanks to report it and for the very well documented issue description!

**Comment by jgrocha** [comment]
Many thanks @vcloarec for the prompt reply. I've tested your PR and it solves the issue. Please keep your good work on QGIS!
